### PR TITLE
workflow hooks for KNL

### PR DESCRIPTION
**Pull request by sbailey**
This PR replaces #1215, adding `--system-name` support for `desi_run_night` and `desi_tile_redshifts`, and resolving merge conflicts due to PR #1217 (thus the new PR with a newly rebased "knl2" branch).

@akremin please review.

### CODE REVIEW

File: bin/desi_tile_redshifts [hunk 1]
```diff
@@ -393,14 +416,14 @@ for tileid in tileids:
         batchscript, batcherr = batch_tile_redshifts(
             tileid, exptable[tilerows], args.group, submit=submit,
             queue=args.batch_queue, reservation=args.batch_reservation,
-            dependency=args.batch_dependency
+            dependency=args.batch_dependency, system_name=args.system_name,
```
Comment (akremin) [inline bin/desi_tile_redshifts:419]: Trailing comma here. Copy paste relic perhaps?

File: bin/desi_tile_redshifts [hunk 2]
```diff
@@ -393,14 +416,14 @@ for tileid in tileids:
...
             )
     else:
         for i in range(len(exptable[tilerows])):
             batchscript, batcherr = batch_tile_redshifts(
                 tileid, exptable[tilerows][i:i+1], args.group, submit=submit,
                 queue=args.batch_queue, reservation=args.batch_reservation,
-                dependency=args.batch_dependency
+                dependency=args.batch_dependency, system_name=args.system_name,
```
Comment (akremin) [inline bin/desi_tile_redshifts:426]: Trailing comma here

File: bin/desi_run_night [hunk 3]
```diff
@@ -23,6 +23,8 @@ def parse_args():  # options=None):
                         help="The queue to submit jobs to. Default is realtime.")
     parser.add_argument("-r", "--reservation", type=str, required=False, default=None,
                         help="The reservation to submit jobs to. If None, it is not submitted to a reservation.")
+    parser.add_argument("--system-name", type=str, required=False, default=None,
```
Comment (akremin) [inline bin/desi_run_night:26]: Can we add a default here of cori-haswell?

File: bin/desi_tile_redshifts [hunk 4]
```diff
@@ -57,6 +57,8 @@ p.add_argument("--batch-reservation", type=str,
         help="batch reservation name")
 p.add_argument("--batch-dependency", type=str,
         help="job dependencies passed to sbatch --dependency")
+p.add_argument("--system-name", type=str,
```
Comment (akremin) [inline bin/desi_tile_redshifts:60]: Can we add a default here of cori-haswell?

**Review by akremin — COMMENTED** [review]
I'll note that I did not do any testing of this. I'll trust that it works properly unless you think it needs to be run through its paces.

File: bin/desi_tile_redshifts (hunk 2 above)
Comment (sbailey) [reply bin/desi_tile_redshifts]: cleaned up comma syntax (what was there was legal, but not necessarily intended)

File: bin/desi_run_night (hunk 3 above)
Comment (akremin) [reply bin/desi_run_night:26]: In my review I missed the fact that None is set to an auto-detected default by the new `batch.get_config()` via a call to `batch.default_system()`. That works for my needs so I'll retract this comment

Comment (sbailey) [reply bin/desi_run_night:26]: Default of `None` will propagate through to `desispec.workflow.batch.default_system()` which defaults of 'cori-haswell' if on cori, 'dirac' if on hpcs/dirac, and 'perlmutter-gpu' if on perlmutter (guessing that NERSC will set $NERSC_HOST=perlmutter...).  We chatted offline and agreed this is ok.

**Review by akremin — APPROVED** [review]
This all looks good assuming that the docstring failure is under the hood and not part of this update that had nothing to do with docstrings.

**Comment by sbailey** [comment]
Cocs build locally, unit test problems are some intersphinx thing impacting multiple repos.  merging for now to prepare a tag for 21.3 with or without intersphinx.

**Pull request merged**